### PR TITLE
Add Request controller and Request Router

### LINKED ISSUE #27: BE: Request Controllers
We need an endpoint for the following related to requests: 
\- GET : list of requests for logged in user
\- POST : Create a new request
\- UPDATE  : Update request with approved or decline

### PR DESCRIPTION
**Pull request by B1umenkrAn2**
### What this PR does (required):
Add request controller with 3 endpoint
Add request router 
update app.js for request router

### Any information needed to test this feature (required):
-All endpoints tested with Postman. Maybe will write some unit test late.

### CODE REVIEW

File: server/app.js [hunk 1]
```diff
@@ -45,6 +46,7 @@ app.use((req, res, next) => {
 app.use("/auth", authRouter);
 app.use("/users", userRouter);
 app.use("/profile", profileRouter);
+app.use("/request",requestRouter);
```
Comment (moffatethan) [inline server/app.js]:
```suggestion
app.use("/request", requestRouter);
```

File: server/controllers/request.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
+
+
```
Comment (moffatethan) [inline server/controllers/request.js]: Unnecessary whitespace here.

File: server/controllers/request.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+   
+const Requset = require("../models/Request");
+const asyncHandler = require("express-async-handler");
+
+// @route GET /request/
+// @desc Get this user all request 
+// @access Private
+exports.getAllRequest = asyncHandler(async (req, res, next) => {
+    const userId  = req.user.id
+    const requests = await Request.find({userId:userId})
```
Comment (moffatethan) [inline server/controllers/request.js]:
```suggestion
    const requests = await Request.find({userId: userId})
```

File: server/controllers/request.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    res.status(200).json({success:{requests}})
```
Comment (moffatethan) [inline server/controllers/request.js]: Not sure why we are squishing everything together. Can we try going forward to space things a part for more readability? Thanks!

Also, `success` property name doesn't quite make sense. We can just do `{ requests }` and that will be fine.

File: server/controllers/request.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+})
+
+// @route POST /request/
+// @desc create new request 
+// @access Private
+exports.createRequest = asyncHandler(async (req, res, next) => {
+
+  const {sitterId,start,end,description} = req.body
```
Comment (moffatethan) [inline server/controllers/request.js]: Please have proper spacing in your destructured object.

File: server/controllers/request.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  const userId = req.user.id
+
+  const request = await Request.create({userId,sitterId,start,end,description})
```
Comment (moffatethan) [inline server/controllers/request.js]: Ditto!

File: server/controllers/request.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+  if(request){
+    res.status(201).json({success:{request}})
```
Comment (moffatethan) [inline server/controllers/request.js]: Similar thoughts to my comment above.

File: server/controllers/request.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+// @access private
+exports.updateRequestById = asyncHandler(async (req, res, next) => {
+  
+  const {accepted} = req.body
+  
+  const request = await Request.findById(req.params.id)
+
+  if(request){
+    
+    // if accepted is fasle, it means that sitter decline this request
```
Comment (moffatethan) [inline server/controllers/request.js]: Code is clear on how it works, we can remove this comment 😄

File: server/controllers/request.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    if(accepted){
+      request.accepted = accepted
+    }else{
+      request.declined = !accepted
+    }
```
Comment (moffatethan) [inline server/controllers/request.js]: This conditional is unusual. Can we make this more intuitive? I mean technically we could just update two properties but I think what would be even better is also taking in a `declined` boolean in the request and then we can use something like `findByIdAndUpdate` instead.

Comment (moffatethan) [reply server/controllers/request.js]: That would get rid of this whole conditional! We should bring in middleware validation as well.

File: server/controllers/request.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    const updateRequest = await request.save()
+    res.status(200).json({success:{request}})
+
+  }else{
+    res.status(400)
```
Comment (moffatethan) [inline server/controllers/request.js]: 404 is the proper status code.

File: server/models/Request.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
+const mongoose = require("mongoose");
```
Comment (moffatethan) [inline server/models/Request.js:1]: Please base your PR branch to the model one so this file is not included in this PR. Thanks!

File: server/routes/request.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+const express = require("express");
+const router = express.Router();
+const protect = require("../middleware/auth");
+const { getAllRequest,createRequest,updateRequestById } = require("../controllers/request");
+
+router.route("/").get(protect, getAllRequest).post(protect,createRequest);
```
Comment (moffatethan) [inline server/routes/request.js]: Proper whitespace please!

**Review by moffatethan — CHANGES_REQUESTED** [review]
Overall decent work, please touch up the code a bit more. A lot of this is just style, let's be mindful of our code style in the codebase. It makes things messy and hard to read, before submitting any PR please go over your code and fix these issues so they don't show up in future PRs. Thanks!

Please also be a bit more descriptive in your PR descriptions. They are very simple and don't provide a lot of information to the reviewer. You should get in the habit of doing more proper PR descriptions showing change logs and notes you want the reviewer to be aware of. This will help you in the long-run so a good time to start practicing is now.

File: server/controllers/request.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
+
+   
+const Requset = require("../models/Request");
+const asyncHandler = require("express-async-handler");
+
+// @route GET /request/
+// @desc Get this user all request 
+// @access Private
```
Comment (moffatethan) [inline server/controllers/request.js:6]: Good job on the documentation!

**Review by B1umenkrAn2 — COMMENTED** [review]
update for formating and typing error

File: server/models/Request.js [hunk 14]
```diff
@@ -13,29 +13,29 @@ const requestSchema = new mongoose.Schema({
   },
   start: {
     type: Date,
-    required:true
+    required: true
```
Comment (moffatethan) [inline server/models/Request.js]: @B1umenkrAn2 This file is conflicting with the main branch now that you've merged in the model. Please address the conflict

File: server/controllers/request.js (hunk 10 above)
Comment (moffatethan) [reply server/controllers/request.js]: @B1umenkrAn2 Seems this one slipped through the cracks. Proper status code is 404 for this. Please change this

**Review by moffatethan — CHANGES_REQUESTED** [review]
Conflict needs to be addressed & you missed a comment I had on the status code change. Style looks better though.

File: server/controllers/request.js [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+})
+
+// @route PUT /request/:id/edit
+// @desc update request by id
+// @access private
+exports.updateRequestById = asyncHandler(async (req, res, next) => {
+  const { accepted, declined } = req.body
+
+  let request = await Request.findOneAndUpdate(req.params.id, { accepted, declined }, { new: true })
```
Comment (moffatethan) [inline server/controllers/request.js]: `const` instead of `let` please

**Review by moffatethan — CHANGES_REQUESTED** [review]

File: server/controllers/request.js (hunk 10 above)
Comment (ugur-akin) [reply server/controllers/request.js]: Hey @B1umenkrAn2, please resolve this before we can go ahead and approve this pull request! 🙏🏼

**Review by moffatethan — APPROVED** [review]
